openassessment/xblock/test/test_grade.py

```diff
@@ -320,28 +320,38 @@ def test_grade_waiting(self, xblock, data):
         # Verify that we're on the waiting template
         self.assertIn(data["expected_response"], resp.decode('utf-8').lower())
 
+    @ddt.data('self', 'peer', 'both')
     @scenario('data/grade_incomplete_scenario.xml', user_id='Bunk')
-    def test_grade_incomplete_missing_self(self, xblock):
-        # Graded peers, but haven't completed self assessment
-        self.create_submission_and_assessments(
-            xblock, self.SUBMISSION, [self.PEERS[0]], [PEER_ASSESSMENTS[0]], None
-        )
-        resp = self.request(xblock, 'render_grade', json.dumps(dict()))
-
-        # Verify that we're on the right template
-        self.assertIn(u'not completed', resp.decode('utf-8').lower())
+    def test_grade_incomplete_missing(self, xblock, missing_step):
+        self_assessment = None
+        peers = []
+        peer_assessments = []
+        if missing_step == 'self':
+            # Graded peers, but haven't completed self assessment
+            peers = [self.PEERS[0]]
+            peer_assessments = [PEER_ASSESSMENTS[0]]
```

Comment:
If we have to do specialized logic, I think it would make more sense to make a separate test function for each one with shared set up and/or assertions instead of trying to work around `ddt`.